File: contest/ICPC/2018/欧洲/NWERC西北欧/data/fastestspeedrun/submissions/time_limit_exceeded/rgl-n3.py
#!/usr/bin/env python3

res = 0

n = int(input()) + 1
edge = [[i == j for j in range(n)] for i in range(n)]
cost = [[     0 for j in range(n)] for i in range(n)]
for i in range(1, n):
  row = list(map(int, input().split()))
  x,t = row[0],row[1]
  res += t
  edge[x][i] = True
  for j in range(n):
    cost[j][i] = row[2+j] - t

for i in range(n):
  for j in range(n):
    for k in range(n):
      edge[j][k] = edge[j][k] or edge[j][i] and edge[i][k]

for i in range(n):
  for j in range(n):
    if edge[i][j]:
      cost[i][j] = 0
      for k in range(n):
        cost[i][k] = min(cost[i][k], cost[j][k])

for i in range(1,n):
  if min([not edge[j][i] or edge[i][j] and (cost[n-1][i],i) <= (cost[n-1][j],j) for j in range(n)]):
    rcost = cost[n-1][i]
    res += rcost
    for j in range(n):
      if edge[i][j] and edge[j][i]:
        for k in range(n):
          cost[k][j] -= rcost

for i in range(n):
  for j in range(n):
    for k in range(n):
      cost[j][k] = min(cost[j][k], max(0,cost[j][i]) + max(0,cost[i][k]))

print(cost[0][-1]+res)

File: contest/ICPC/2018/欧洲/NWERC西北欧/data/fastestspeedrun/submissions/time_limit_exceeded/rgl-n3.cc
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

bool edge[5001][5001];
ll cost[5001][5001];

#define f(i) for (int i=n; i--;)
#define f2(i,j) f(i) f(j)
#define f3(i,j,k) f(i) f(j) f(k)

int main(){
  int n; cin>>n; ++n;
  ll res=0;

  for (int i=1; i<n; i++){
    int x; ll t; cin>>x>>t; res+=t;
    edge[i][i]=edge[x][i]=true;
    for (int j=0; j<n; j++){
      cin>>cost[j][i]; cost[j][i]-=t;
    }
  }

  f3(i,j,k) edge[j][k]|=edge[j][i]&edge[i][k];
  f2(i,j) if (edge[i][j]) f(k) cost[i][k]=min(cost[i][k],cost[j][k]);
  f2(i,j) if (edge[i][j]) cost[i][j]=0;

  for (int i=1; i<n; i++){
    bool is_root=true;
    f(j) is_root&=edge[i][j] and make_pair(cost[n-1][i],i)<=make_pair(cost[n-1][j],j) or not edge[j][i];
    if (is_root){
      auto const rcost=cost[n-1][i]; res+=rcost;
      f2(j,k) if (edge[i][j] and edge[j][i]) cost[k][j]-=rcost;
    }
  }

  f2(i,j) if (edge[i][j]) cost[i][j]=0;
  f3(i,j,k) cost[j][k]=min(cost[j][k],cost[j][i]+cost[i][k]);
  cout<<cost[0][n-1]+res<<endl;
}
